File: 206_reverse_linked_list.cpp
/**
 * 206. 反转链表
 * https://leetcode.com/problems/reverse-linked-list/
 */

#include <iostream>
#include <vector>
#include <memory>
using namespace std;

/**
 * 链表节点定义
 */
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
    /**
     * 解法一：迭代法（双指针）
     */
    ListNode* reverseList_iterative(ListNode* head) {
        /**
         * 迭代法 - 双指针
         * 时间复杂度: O(n)，n 为链表长度
         * 空间复杂度: O(1)
         */
        ListNode* prev = nullptr;  // 前一个节点（反转后链表的头）
        ListNode* curr = head;     // 当前节点

        while (curr) {
            // 1. 保存下一个节点
            ListNode* next_temp = curr->next;

            // 2. 反转指针
            curr->next = prev;

            // 3. 移动指针
            prev = curr;
            curr = next_temp;
        }

        // prev 最终指向新的头节点
        return prev;
    }

    /**
     * 解法二：递归法
     */
    ListNode* reverseList_recursive(ListNode* head) {
        /**
         * 递归法
         * 时间复杂度: O(n)，n 为链表长度
         * 空间复杂度: O(n)，递归栈深度
         */
        // 基准条件：空链表或只有一个节点
        if (!head || !head->next) {
            return head;
        }

        // 递归处理子链表
        ListNode* new_head = reverseList_recursive(head->next);

        // 反转指针
        head->next->next = head;
        head->next = nullptr;

        return new_head;
    }
};

/**
 * 辅助函数
 */
ListNode* createLinkedList(const vector<int>& values) {
    if (values.empty()) {
        return nullptr;
    }

    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    for (size_t i = 1; i < values.size(); i++) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }

    return head;
}

vector<int> linkedListToVector(ListNode* head) {
    vector<int> result;
    ListNode* current = head;
    while (current) {
        result.push_back(current->val);
        current = current->next;
    }
    return result;
}

bool compareLinkedLists(ListNode* l1, ListNode* l2) {
    while (l1 && l2) {
        if (l1->val != l2->val) {
            return false;
        }
        l1 = l1->next;
        l2 = l2->next;
    }
    return l1 == nullptr && l2 == nullptr;
}

void deleteLinkedList(ListNode* head) {
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

/**
 * 测试函数
 */
void testSolution(const string& name, ListNode* (Solution::*method)(ListNode*)) {
    cout << "\n" << string(60, '=') << "\n";
    cout << "测试方法: " << name << "\n";
    cout << string(60, '=') << "\n";

    vector<vector<int>> test_cases = {
        {1, 2, 3, 4, 5},     // 示例1：5个节点
        {1, 2},              // 示例2：2个节点
        {},                  // 示例3：空链表
        {1},                 // 单个节点
        {1, 2, 3},           // 3个节点
        {1, 2, 3, 4},        // 4个节点
        {1, 2, 3, 4, 5, 6, 7, 8, 9, 10},  // 10个节点
        {0, 0, 0, 0},        // 重复值
        {-1, -2, -3, -4},    // 负数
        {5, 3, 1, 2, 4}      // 乱序数组
    };

    vector<string> descriptions = {
        "示例1：5个节点",
        "示例2：2个节点",
        "示例3：空链表",
        "单个节点",
        "3个节点",
        "4个节点",
        "10个节点",
        "重复值",
        "负数",
        "乱序数组"
    };

    int passed = 0;
    int failed = 0;

    for (size_t i = 0; i < test_cases.size(); i++) {
        const auto& input = test_cases[i];
        const string& description = descriptions[i];

        // 计算期望输出（反转）
        vector<int> expected = input;
        reverse(expected.begin(), expected.end());

        try {
            // 创建输入链表
            ListNode* head = createLinkedList(input);

            // 执行反转
            Solution solution;
            ListNode* result_head = (solution.*method)(head);

            // 转换为向量
            vector<int> result = linkedListToVector(result_head);

            // 验证结果
            if (result == expected) {
                cout << "✓ 测试 " << (i + 1) << " 通过: " << description << "\n";
                cout << "  输入: [";
                for (size_t j = 0; j < input.size(); j++) {
                    cout << input[j];
                    if (j < input.size() - 1) cout << ", ";
                }
                cout << "]\n";
                cout << "  输出: [";
                for (size_t j = 0; j < result.size(); j++) {
                    cout << result[j];
                    if (j < result.size() - 1) cout << ", ";
                }
                cout << "]\n";
                passed++;
            } else {
                cout << "✗ 测试 " << (i + 1) << " 失败: " << description << "\n";
                cout << "  输入: [";
                for (size_t j = 0; j < input.size(); j++) {
                    cout << input[j];
                    if (j < input.size() - 1) cout << ", ";
                }
                cout << "]\n";
                cout << "  期望: [";
                for (size_t j = 0; j < expected.size(); j++) {
                    cout << expected[j];
                    if (j < expected.size() - 1) cout << ", ";
                }
                cout << "]\n";
                cout << "  实际: [";
                for (size_t j = 0; j < result.size(); j++) {
                    cout << result[j];
                    if (j < result.size() - 1) cout << ", ";
                }
                cout << "]\n";
                failed++;
            }

            // 清理内存
            deleteLinkedList(result_head);

        } catch (const exception& e) {
            cout << "✗ 测试 " << (i + 1) << " 出现异常: " << description << "\n";
            cout << "  异常: " << e.what() << "\n";
            failed++;
        }
    }

    cout << string(60, '-') << "\n";
    cout << "测试结果: " << passed << " 通过, " << failed << " 失败\n";
    cout << string(60, '-') << "\n";

    if (failed == 0) {
        cout << "✓ " << name << " 所有测试通过!\n";
    }
}

int main() {
    cout << "=" << string(60, '=') << "\n";
    cout << "测试反转链表\n";
    cout << string(60, '=') << "\n";

    // 测试迭代法
    testSolution("迭代法 (双指针)", &Solution::reverseList_iterative);

    // 测试递归法
    testSolution("递归法", &Solution::reverseList_recursive);

    cout << "\n" << string(60, '=') << "\n";
    cout << "所有测试完成!\n";
    cout << string(60, '=') << "\n";

    return 0;
}

File: 206_reverse_linked_list.py
"""
206. 反转链表
https://leetcode.com/problems/reverse-linked-list/
"""

from typing import Optional, List


class ListNode:
    """链表节点"""
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

    def __repr__(self):
        return f"ListNode({self.val})"

    def to_list(self) -> List[int]:
        """将链表转换为列表"""
        result = []
        node = self
        while node:
            result.append(node.val)
            node = node.next
        return result


class Solution:
    """解法一：迭代法（双指针）"""
    def reverseList_iterative(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        迭代法 - 双指针
        时间复杂度: O(n)，n 为链表长度
        空间复杂度: O(1)
        """
        prev = None  # 前一个节点（反转后链表的头）
        curr = head  # 当前节点

        while curr:
            # 1. 保存下一个节点
            next_temp = curr.next

            # 2. 反转指针
            curr.next = prev

            # 3. 移动指针
            prev = curr
            curr = next_temp

        # prev 最终指向新的头节点
        return prev

    """解法二：递归法"""
    def reverseList_recursive(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        递归法
        时间复杂度: O(n)，n 为链表长度
        空间复杂度: O(n)，递归栈深度
        """
        # 基准条件：空链表或只有一个节点
        if not head or not head.next:
            return head

        # 递归处理子链表
        new_head = self.reverseList_recursive(head.next)

        # 反转指针
        head.next.next = head
        head.next = None

        return new_head


def create_linked_list(values: List[int]) -> Optional[ListNode]:
    """根据列表创建链表"""
    if not values:
        return None

    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next

    return head


def compare_linked_lists(l1: Optional[ListNode], l2: Optional[ListNode]) -> bool:
    """比较两个链表是否相等"""
    while l1 and l2:
        if l1.val != l2.val:
            return False
        l1 = l1.next
        l2 = l2.next

    return l1 is None and l2 is None


def test_solution():
    """测试所有解法"""
    print("=" * 60)
    print("测试反转链表")
    print("=" * 60)

    test_cases = [
        {
            "input": [1, 2, 3, 4, 5],
            "expected": [5, 4, 3, 2, 1],
            "description": "示例1：5个节点"
        },
        {
            "input": [1, 2],
            "expected": [2, 1],
            "description": "示例2：2个节点"
        },
        {
            "input": [],
            "expected": [],
            "description": "示例3：空链表"
        },
        {
            "input": [1],
            "expected": [1],
            "description": "单个节点"
        },
        {
            "input": [1, 2, 3],
            "expected": [3, 2, 1],
            "description": "3个节点"
        },
        {
            "input": [1, 2, 3, 4],
            "expected": [4, 3, 2, 1],
            "description": "4个节点"
        },
        {
            "input": list(range(1, 11)),
            "expected": list(range(10, 0, -1)),
            "description": "10个节点"
        },
        {
            "input": [0, 0, 0, 0],
            "expected": [0, 0, 0, 0],
            "description": "重复值"
        },
        {
            "input": [-1, -2, -3, -4],
            "expected": [-4, -3, -2, -1],
            "description": "负数"
        },
        {
            "input": [5, 3, 1, 2, 4],
            "expected": [4, 2, 1, 3, 5],
            "description": "乱序数组"
        }
    ]

    methods = [
        ("迭代法 (双指针)", Solution().reverseList_iterative),
        ("递归法", Solution().reverseList_recursive)
    ]

    for method_name, method in methods:
        print(f"\n{'=' * 60}")
        print(f"测试方法: {method_name}")
        print(f"{'=' * 60}")

        passed = 0
        failed = 0

        for i, test_case in enumerate(test_cases, 1):
            input_list = test_case["input"]
            expected_list = test_case["expected"]
            description = test_case["description"]

            try:
                # 创建输入链表
                head = create_linked_list(input_list)

                # 执行反转
                result_head = method(head)

                # 转换为列表
                result_list = result_head.to_list() if result_head else []

                # 验证结果
                if result_list == expected_list:
                    print(f"✓ 测试 {i} 通过: {description}")
                    print(f"  输入: {input_list}")
                    print(f"  输出: {result_list}")
                    passed += 1
                else:
                    print(f"✗ 测试 {i} 失败: {description}")
                    print(f"  输入: {input_list}")
                    print(f"  期望: {expected_list}")
                    print(f"  实际: {result_list}")
                    failed += 1
            except Exception as e:
                print(f"✗ 测试 {i} 出现异常: {description}")
                print(f"  异常: {e}")
                failed += 1

        print(f"\n{'-' * 60}")
        print(f"测试结果: {passed} 通过, {failed} 失败")
        print(f"{'-' * 60}")

        if failed == 0:
            print(f"✓ {method_name} 所有测试通过!")

    print(f"\n{'=' * 60}")
    print("所有测试完成!")
    print(f"{'=' * 60}")


def test_performance():
    """性能测试"""
    print("\n" + "=" * 60)
    print("性能测试")
    print("=" * 60)

    import time

    # 测试不同规模的链表（递归法对大数据有栈溢出风险）
    sizes = [100, 500, 1000]

    for size in sizes:
        print(f"\n测试规模: {size} 个节点")

        # 创建大链表
        values = list(range(1, size + 1))

        # 测试迭代法
        head = create_linked_list(values)
        start_time = time.time()
        result_head = Solution().reverseList_iterative(head)
        elapsed_time = time.time() - start_time
        print(f"  迭代法: {elapsed_time:.6f} 秒")

        # 测试递归法（仅对小规模测试，Python 默认递归限制约 1000）
        if size <= 800:
            head = create_linked_list(values)
            start_time = time.time()
            result_head = Solution().reverseList_recursive(head)
            elapsed_time = time.time() - start_time
            print(f"  递归法: {elapsed_time:.6f} 秒")
        else:
            print(f"  递归法: 跳过（{size} > 800，Python 递归限制）")


def test_memory_efficiency():
    """内存效率测试"""
    print("\n" + "=" * 60)
    print("内存效率测试")
    print("=" * 60)

    # 测试极端情况：超长链表
    # 递归法可能会导致栈溢出
    try:
        size = 100000
        print(f"\n创建 {size} 个节点的链表...")
        values = list(range(1, size + 1))
        head = create_linked_list(values)

        print("测试迭代法...")
        result = Solution().reverseList_iterative(head)
        print("✓ 迭代法成功处理大链表")

        # 重新创建链表
        head = create_linked_list(values)

        print("测试递归法...")
        result = Solution().reverseList_recursive(head)
        print("✓ 递归法成功处理大链表")

    except RecursionError:
        print("✗ 递归法栈溢出（预期行为）")
    except Exception as e:
        print(f"✗ 测试出现异常: {e}")


if __name__ == "__main__":
    test_solution()
    test_performance()
    test_memory_efficiency()
